index/segment/fst/fst_codec.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fst
+
+import (
+	"bytes"
+
+	"github.com/couchbase/vellum"
+)
+
+// fstCodec is a codec to help construct an FST in memory.
+type fstCodec struct {
+	scratchBuffer bytes.Buffer
+	builder       *vellum.Builder
+}
+
+func newFSTCodec() *fstCodec {
+	return &fstCodec{}
+}
+
+func (f *fstCodec) Clear() {
+	f.scratchBuffer.Reset()
+	f.builder = nil
+}
+
+func (f *fstCodec) Reset() error {
+	f.Clear()
+
+	// TODO(prateek): builderopts for vellum
+	builder, err := vellum.New(&f.scratchBuffer, nil)
+	if err != nil {
+		return err
+	}
+	f.builder = builder
+	return nil
+}
+
+func (f *fstCodec) Add(b []byte, v uint64) error {
+	return f.builder.Insert(b, v)
```

Comment:
Does `vellum.Builder` check if insert was in order or not?  If not should this codec perform that check by comparing against last insert?

Reply:
Yep vellum does: https://github.com/couchbase/vellum/blob/master/builder.go#L94-L96